### ISSUE TITLE
Conv2d interface

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#2505]
We need a new interface. I'll give more detail later today.

**Comment by benanne** [comment]
Yes please! Awesome! Let me know if I can be of use in accomplishing this.

**Comment by f0k** [comment]
> We need a new interface. I'll give more detail later today.

Just linking this discussion on convolution placeholder ops: https://github.com/Theano/Theano/issues/2268#issuecomment-63621626
I'm curious what you'll propose! (And while we're at it, I'm also curious what CCW stands for, actually.)

**Comment by nouiz** [comment]
It is the same, except that a ticket for it is much better then a comment! I updated the description to have the description and a link to that comment.

@benanne We need to find someone to work on it. Do you have the time? If not, until we find someone with the time, I do not see something that you can do to help. When we have a PR, a review from both of you would be great.

**Comment by f0k** [comment]
I don't have the time to implement this right now, but probably in a few weeks. There's an open question to me, though: If the convolution you introduce via `conv2d` becomes a placeholder op without any implementation (that's what I was suggesting), how is the CPU/GPU distinction supposed to work?
Right now `ConvOp` has the CPU implementation, there is a single optimizer replacing `ConvOp` with `GpuConv` (the legacy GPU implementation), and then a bunch of optimizers replacing `GpuConv` with alternative implementations.
In future, `ConvOp` would be a placeholder for the CPU implementation, there would still be a single optimizer replacing `ConvOp` with `GpuConv`, which would be a placeholder for the GPU implementation, and then a bunch of optimizers replacing `GpuConv` with a real implementation that supports whatever is needed. Additionally, we would need an optimizer replacing `ConvOp` with a CPU implementation that supports whatever is needed (right now we only have a single CPU implementation, but we could add the one from caffe). This one must only kick in when the convolution cannot moved to the GPU, though, otherwise it will stop the `ConvOp -> GpuConv` optimizer from working.
Do you follow me? What would be the best solution?

**Comment by lamblin** [comment]
@f0k I think what you are proposing makes sense. In order:
- CPU placeholder -> GPU placeholder
- GPU placeholder -> GPU implementation
- CPU placeholder -> CPU implementation
CCW stands for "common code workflow", which is how we call the process by which we organize the contribution from our lab's students to the common software tools we use.

**Comment by nouiz** [comment]
I think we can reuse the same placeholder for CPU and GPU, and only compare
the input type to make the difference. This lower the number of class.

**Comment by f0k** [comment]
@lamblin: If you can prioritize groups of optimizers in this way, this sounds very clean.
@nouiz: I'd personally think that having an explicit GPU placeholder would make it easier to write/understand the optimizers. But unless I'm missing something, the GPU placeholder could be derived from the CPU placeholder without introducing any code on its own.

Next question: Will we have separate placeholders for the forward pass and the two gradients, or will we have a placeholder that can express all three? (I now tend towards the former. This will probably lead to less headaches in the placeholder -> implementation optimizers, even if it may lead to more code.)

Related question: Which ops will have a `grad()` method? Theoretically, it would be enough if only the CPU placeholder(s) had it. But we already implemented them for the gemm-based convolution and the cudnn-based convolution, and should keep them at the very least for backwards compatibility. We don't necessarily need them for the fft-based convolution and the legacy implementation.

**Comment by nouiz** [comment]
Agree on all except I would keep the grad on the legacy op.
I would also go for 3 op, but one with a direction is also good. I think 3
op make thing more clear.

**Comment by f0k** [comment]
> Agree on all except I would keep the grad on the legacy op.

The legacy grad is quite terrible, though.

> I would also go for 3 op, but one with a direction is also good. I think 3
op make thing more clear.

I wasn't actually thinking of a single op with a direction parameter, but a single op that can express all the operations that can happen (which is easy if you don't have strides and only "full" and "valid" convolutions, but becomes difficult with strides and custom padding). This would make things more complicated, though.

One op with a direction parameter is an interesting idea as well, it would reduce code duplication for some optimizers. But I also think splitting it makes things easier to follow, especially as the gradient ops will need the symbolic input or kernel shape as additional input variables.

OK, last question: What should the interface look like? Proposal:
```python
def conv2d(input, filters, input_shape=None, filter_shape=None, border_mode='valid',
        subsample=(1, 1), filter_flip=True)
    """
    This function will build the symbolic graph for convolving a mini-batch of a
    stack of 2D inputs with a set of 2D filters. The implementation is modelled
    after Convolutional Neural Networks (CNN).

    :type input: symbolic 4D tensor
    :param input: mini-batch of feature map stacks, of shape
        (batch size, input channels, input rows, input columns).
        See the optional parameter ``input_shape``.

    :type filters: symbolic 4D tensor
    :param filters: set of filters used in CNN layer of shape
        (output channels, input channels, filter rows, filter columns).
        See the optional parameter ``filter_shape``.

    :type input_shape: None, tuple/list of len 4 of int or Constant variable
    :param input_shape: The shape of the input parameter.
        Optional, possibly used to choose an optimal implementation.
        You can give ``None`` for any element of the list to specify that this
        element is not known at compile time.

    :type filter_shape: None, tuple/list of len 4 of int or Constant variable
    :param filter_shape: The shape of the filters parameter.
        Optional, possibly used to choose an optimal implementation.
        You can give ``None`` for any element of the list to specify that this
        element is not known at compile time.

    :type border_mode: str, int or tuple of two int
    :param border_mode: Either of the following:
        * ``'valid'``: apply filter wherever it completely overlaps with the
          input. Generates output of shape: input shape - filter shape + 1
        * ``'full'``: apply filter wherever it partly overlaps with the input.
          Generates output of shape: input shape + filter shape - 1
        * ``'half'``: pad input with a symmetric border of ``filter rows // 2``
          rows and ``filter columns // 2`` columns, then perform a valid
          convolution. For filters with an odd number of rows and columns, this
          leads to the output shape being equal to the input shape.
        * ``int``: pad input with a symmetric border of zeros of the given
          width, then perform a valid convolution.
        * ``(int1, int2)``: pad input with a symmetric border of ``int1`` rows
          and ``int2`` columns, then perform a valid convolution.

    :type subsample: tuple of len 2
    :param subsample: factor by which to subsample the output.
        Also called strides elsewhere.
    
    :type filter_flip: bool
    :param filter_flip: If ``True``, will flip the filter rows and columns
        before sliding them over the input. This opearation is normally referred
        to as a convolution, and this is the default. If ``False``, the filters
        are not flipped and the operation is referred to as a cross-correlation.

    :rtype: symbolic 4D tensor
    :return: set of feature maps generated by convolutional layer. Tensor is
        of shape (batch size, output channels, output rows, output columns)
    """
```

**Comment by nouiz** [comment]
It need a pad parameter. Instead of filter_flip, why not reuse
conv_mode='conv' or 'corr' as in dnn_conv?

**Comment by f0k** [comment]
> It need a pad parameter.

Custom padding has been integrated in `border_mode`. I thought we had decided on that earlier?

> Instead of filter_flip, why not reuse conv_mode='conv' or 'corr' as in dnn_conv?

Okay, fine by me (although setting "convolution" to "correlation" sounds a bit strange... but it's the same in the original cuDNN API).

**Comment by nouiz** [comment]
I forgot that padding was included in the border_mode. That is fine.

I do not understand your last sentence. conv correspond to convolution and
this do the kernel flipping. corr, correspond to cross-correlation and it
do not do the kernel flipping.

**Comment by f0k** [comment]
> I do not understand your last sentence.

It's just the name of the parameter. `conv_mode="corr"` or `convolution_mode="correlation"` sounds about as nonsensical as `corr_mode="conv"` or `correlation_mode="convolution"`. That's why I came up with the more neutral `filter_flip`. I don't have a strong opinion on that, though, let's just go with the name that's already established in `conv_dnn`.

**Comment by nouiz** [comment]
I see. I like the idea to change the name. We do not need to follow the
cudnn convention in the new generic op as it is the only one that offer the
choice.

What other people think about this? It there other convolution mode that
exist? I do not know others.

**Comment by lamblin** [comment]
Answers to various points:
- Having the different optimizers run in different phases should take care of the ordering/priority
- I'm favoring 3 Ops, otherwise we would be forced to support an interface much more complex (subsampling or supersampling independently on each input and output), and derive the gradients for all those cases which do not have any implementation.
- The `grad` method of the legacy implementation should build a placeholder Op.
- Do we need a CPU implementation of the two gradient Ops in the case of subsampling right now? Or should the placeholder simply be replaced by the current graph using Conv3D?

**Comment by nouiz** [comment]
I didn't taught of point 3, but I agree.

I think the opt should replace with the current graph for now.
Making/porting a new implementation on the CPU for that case is not in this
issue.

**Comment by f0k** [comment]
Good, we're converging on three optimization phases and three ops.

> The `grad` method of the legacy implementation should build a placeholder Op.

Sounds elegant, although it will prevent people from introducing a pure-legacy convolution somewhere (which is currently possible for cudnn and caffe without disabling any optimizers). I wouldn't mind, though.

> Do we need a CPU implementation of the two gradient Ops in the case of subsampling right now? Or should the placeholder simply be replaced by the current graph using Conv3D?

I guess we can just use the Conv3D fallback for now. This can be changed at any time later.

**Comment by FredericGodin** [comment]
Is there any news on the 1D convolution?

**Comment by nouiz** [comment]
It isn't ready. We will finish the 2d first. There is code around that
allow to use conv2d to make 1d conv that use the "best" order to make it
fast.

Fred

**Comment by nouiz** [comment]
I think we are only missing conv1d. I'll add that as a summary in the description and remove the milestone.
